### ISSUE TITLE
ArangoDB crashes / does no longer respond on Ubuntu 14.04

### ISSUE THREAD

**Issue by safebyte** [arangodb/arangodb#2218]
Good morning.

I've just been woken up by a colleague because ArangoDB stoped responding _again_ for the 3rd time in 4 days. Unfortunatley without any meaningful logs - neither in `arangod.log` nor in `arangod.log.supervisor`. Also I'm unable to reproduce the crash; it just happens approx. every 12 hours to 3 days. 

<sup><sub>_Please don't make me switch back to mySQL, I have nightmares about it - but at least it doesn't crash or stop respoding completly._</sub></sup>

## my environment running ArangoDB
I'm using ArangoDB:
- [x] 3.1.3 64bit

On this operating system:
- [x] Linux 
  - [x] Ubuntu 14.04.5

#### What happend
- Our FOXX App does accept requests but they run for ever without sending any response
- Same for the Web Interface
- `service arangodb3 status` states that ArangoDB is (still) running
- `service arangodb3 restart` does no longer work - it freezes
- The process has to be killed manually to get it running again

#### My `arangod.conf` (modified parts):
    endpoint = ssl://www.mytld.de:6969
 
    [ssl]
    keyfile = /valid/path/to/the/cert/tld_de.pem
    cafile = /valid/path/to/the/cert/tld_de.ca-bundle

#### About our setup:
- We're running ArangoDB on a single server (no replication)
- We're using our own FOXX Service (on production mode)
- Our Backend does a few queries to ArangoDB, but the FOXX app does most of the work
- We have very little data: 7 collections with a maximum of 200 items per collection
- The Web Interface states a memory usage of 638.66 MB (4.06 %) (after restart)
- We're using the `_system` database for our collections
- I have never experienced the issue (using the same data) on my development environment (Win10) - maybe because it never runs long enough

#### Symptoms:
The following logs might be an indication of what is happening:

    2016-12-13T03:35:00Z [1514] ERROR {communication} unable to perform ssl handshake: wrong version number : 336130315
    2016-12-13T03:44:53Z [1514] ERROR {communication} unable to perform ssl handshake: short read : 335544539

- The first line is repeated _randomly_ between 5 to 20 times every now and then
- None of these logs lead to the described behavour directly (keeps running).
- There are no logs for the crash at all

#### Startup logs:

    2016-12-13T02:29:34Z [1116] INFO ArangoDB 3.1.3 [linux] 64bit, using VPack 0.1.30, ICU 54.1, V8 5.0.71.39, OpenSSL 1.0.1f 6 Jan 2014
    2016-12-13T02:29:34Z [1116] INFO file-descriptors (nofiles) hard limit is 131072, soft limit is 131072
    2016-12-13T02:29:37Z [1116] INFO running WAL recovery (1 logfiles)
    2016-12-13T02:29:37Z [1116] INFO replaying WAL logfile '/var/lib/arangodb3/journals/logfile-1872941.db' (1 of 1)
    2016-12-13T02:29:37Z [1116] INFO WAL recovery finished successfully
    2016-12-13T02:29:37Z [1116] INFO JavaScript using startup '/usr/share/arangodb3/js', application '/var/lib/arangodb3-apps'
    2016-12-13T02:29:39Z [1514] INFO ArangoDB 3.1.3 [linux] 64bit, using VPack 0.1.30, ICU 54.1, V8 5.0.71.39, OpenSSL 1.0.1f 6 Jan 2014
    2016-12-13T02:29:39Z [1514] INFO using SSL options: SSL_OP_CIPHER_SERVER_PREFERENCE, SSL_OP_TLS_ROLLBACK_BUG
    2016-12-13T02:29:39Z [1514] INFO Authentication is turned on
    2016-12-13T02:29:39Z [1514] INFO Authentication system only
    2016-12-13T02:29:39Z [1514] INFO Authentication for unix sockets is turned on
    2016-12-13T02:29:39Z [1514] INFO file-descriptors (nofiles) hard limit is 131072, soft limit is 131072
    2016-12-13T02:29:39Z [1514] INFO JavaScript using startup '/usr/share/arangodb3/js', application '/var/lib/arangodb3-apps'
    2016-12-13T02:29:40Z [1514] INFO In database '_system': Database is up-to-date (30103/standalone/existing)
    2016-12-13T02:29:40Z [1514] INFO using endpoint 'http+ssl://www.mytld.de:6969' for ssl-encrypted requests
    2016-12-13T02:29:41Z [1514] INFO ArangoDB (version 3.1.3 [linux]) is ready for business. Have fun!

I'm running out of ideas to debug this - any help is very much appreciated because I need my sleep and this is on a production system.

Thank you

**Comment by jsteemann** [comment]
First of all, could you upgrade to 3.1.4? It is the latest 3.1 package released and contains some bugfixes (that may or may not be related to the issue).
In the meantime, we'll try to create an arangod binary with debug symbols that can be used for analyzing the problem. When the problem re-occurs, attaching to this binary with gdb and producing a backtrace may reveal the problem source.

**Comment by safebyte** [comment]
We've now upgraded to 3.1.4. Should I look for something specific in case it freezes before we have the binary? Thanks

**Comment by safebyte** [comment]
PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
     7732 arangodb  20   0 1215992 364524 167268 R 196.8  2.3  12:13.98 arangodb [s+

The arango process runs at almost 200% cpu right now. The Foxx app is still answering but a lot slower than usual. The ArangoDB Web Interface does no longer respond. I've enabled the trace log earier, maybe it helps:


    ...
    2016-12-14T15:19:54Z [8163] TRACE returned dirty V8 context
    2016-12-14T15:19:54Z [8163] TRACE V8 context has reached GC timeout threshold and will be scheduled for GC
    2016-12-14T15:19:55Z [8163] TRACE found unused V8 context
    2016-12-14T15:19:55Z [8163] TRACE entering V8 context 0
    2016-12-14T15:19:55Z [8163] TRACE leaving V8 context 0
    2016-12-14T15:19:55Z [8163] TRACE returned dirty V8 context
    2016-12-14T15:19:55Z [8163] TRACE V8 context has reached GC timeout threshold and will be scheduled for GC
    2016-12-14T15:19:56Z [8163] TRACE found unused V8 context
    2016-12-14T15:19:56Z [8163] TRACE entering V8 context 0
    2016-12-14T15:19:56Z [8163] TRACE leaving V8 context 0
    2016-12-14T15:19:56Z [8163] TRACE returned dirty V8 context
    2016-12-14T15:19:56Z [8163] TRACE V8 context has reached GC timeout threshold and will be scheduled for GC
    2016-12-14T15:19:57Z [8163] TRACE found unused V8 context
    2016-12-14T15:19:57Z [8163] TRACE entering V8 context 0
    2016-12-14T15:19:57Z [8163] TRACE leaving V8 context 0
    2016-12-14T15:19:57Z [8163] TRACE returned dirty V8 context
    2016-12-14T15:19:57Z [8163] TRACE V8 context has reached GC timeout threshold and will be scheduled for GC
    2016-12-14T15:19:58Z [8163] TRACE found unused V8 context
    2016-12-14T15:19:58Z [8163] TRACE entering V8 context 0
    2016-12-14T15:19:58Z [8163] TRACE leaving V8 context 0
    ... and so on ...

**Comment by safebyte** [comment]
I narrowed it down to a Foxx SSL issue and switched all APIs to http (to have them running at least).

#### The following lead to the crash:
- SSL endpoint configured as in the first post
- Creating an empty Foxx app in the Web Interface
- Sending a simple GET request to the namespace of the Foxx APP by using the HTTPS interface
- API responds properly / and even keeps responding to the same connection
- But all other client connections time out
- The arangod process has ~100% cpu
- And even though a `service arangodb3 stop` finishes with a success message
- The process has to be manually killed to be restarted

As mentioned before, it seems to only affect our production system (Ubuntu 14).

**Comment by jsteemann** [comment]
Thanks for the details. We'll try reproducing this with an SSL setup. 
Is there a network in between the client application(s) and the ArangoDB server hosting the Foxx services or are client applications and ArangoDB running on the same host?

**Comment by jsteemann** [comment]
Got it into a state in which one Scheduler thread uses about 100% CPU:
```
  PID USER      PR  NI    VIRT    RES    SHR S %CPU %MEM     TIME+ COMMAND      
13787 steemann  20   0 5970300 524656 125076 R 99,9  3,2   0:54.13 Scheduler    
```
Backtrace for the thread:
```
#0  0x00007f7eea3c799d in recvmsg () at ../sysdeps/unix/syscall-template.S:84
#1  0x0000000000afa8eb in boost::asio::detail::socket_ops::recv (ec=..., flags=0, count=1, bufs=0x7f7ed7ffb620, s=30)
    at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/detail/impl/socket_ops.ipp:784
#2  boost::asio::detail::socket_ops::sync_recv (ec=..., all_empty=<optimized out>, flags=0, count=1, bufs=0x7f7ed7ffb620, state=<optimized out>, 
    s=30) at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/detail/impl/socket_ops.ipp:811
#3  boost::asio::detail::reactive_socket_service_base::receive<boost::asio::mutable_buffers_1> (this=<optimized out>, impl=..., impl=..., 
    ec=..., flags=0, buffers=...) at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/detail/reactive_socket_service_base.hpp:255
#4  boost::asio::stream_socket_service<boost::asio::ip::tcp>::receive<boost::asio::mutable_buffers_1> (flags=0, this=<optimized out>, ec=..., 
    buffers=..., impl=...) at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/stream_socket_service.hpp:341
#5  boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> >::read_some<boost::asio::mutable_buffers_1> (ec=..., buffers=..., this=0x7f7ecc8a37b0)
    at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/basic_stream_socket.hpp:795
#6  boost::asio::ssl::detail::io<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> >, boost::asio::ssl::detail::handshake_op> (ec=..., op=..., core=..., next_layer=...)
    at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/ssl/detail/io.hpp:47
#7  boost::asio::ssl::stream<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> > >::handshake (ec=..., type=boost::asio::ssl::stream_base::server, this=0x7f7ecc8a37b0)
    at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/ssl/stream.hpp:381
#8  arangodb::socketcommon::doSslHandshake<boost::asio::ssl::stream<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> > > > (socket=...) at /home/steemann/Arango31NoAsan/arangod/Scheduler/Socket.h:49
#9  0x0000000000af2f13 in arangodb::rest::SocketTask::SocketTask(arangodb::EventLoop, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >, arangodb::ConnectionInfo&&, double, bool) (this=0x7f7edc2a2590, __vtt_parm=<optimized out>, loop=..., 
    socket=std::unique_ptr<arangodb::Socket> containing 0x7f7ed7ffb960, connectionInfo=<optimized out>, keepAliveTimeout=300, skipInit=false, 
    __in_chrg=<optimized out>) at /home/steemann/Arango31NoAsan/arangod/Scheduler/SocketTask.cpp:61
#10 0x000000000096ec0e in arangodb::rest::GeneralCommTask::GeneralCommTask(arangodb::EventLoop, arangodb::rest::GeneralServer*, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >, arangodb::ConnectionInfo&&, double, bool) (this=0x7f7edc2a2590, 
    __vtt_parm=0x18be450 <VTT for arangodb::rest::HttpCommTask+8>, loop=..., server=0x7f7ed42a58b0, socket=..., 
    info=<unknown type in /home/steemann/Arango31NoAsan/build/bin/arangod, CU 0x39ff7c6, DIE 0x3a64354>, keepAliveTimeout=<optimized out>, 
    skipSocketInit=false, __in_chrg=<optimized out>) at /home/steemann/Arango31NoAsan/arangod/GeneralServer/GeneralCommTask.cpp:60
#11 0x000000000097e9ff in arangodb::rest::HttpCommTask::HttpCommTask(arangodb::EventLoop, arangodb::rest::GeneralServer*, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >, arangodb::ConnectionInfo&&, double) (this=0x7f7edc2a2590, loop=..., server=0x7f7ed42a58b0, 
    socket=std::unique_ptr<arangodb::Socket> containing 0x7f7ed7ffba70, 
    info=<unknown type in /home/steemann/Arango31NoAsan/build/bin/arangod, CU 0x3bfdfb2, DIE 0x3c850c8>, timeout=300, __in_chrg=<optimized out>, 
    __vtt_parm=<optimized out>) at /home/steemann/Arango31NoAsan/arangod/GeneralServer/HttpCommTask.cpp:65
#12 0x0000000000975b70 in __gnu_cxx::new_allocator<arangodb::rest::HttpCommTask>::construct<arangodb::rest::HttpCommTask, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >, arangodb::ConnectionInfo, double&>(arangodb::rest::HttpCommTask*, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >&&, arangodb::ConnectionInfo&&, double&) (__p=0x7f7edc2a2590, this=<optimized out>) at /usr/include/c++/5/ext/new_allocator.h:120
#13 std::allocator_traits<std::allocator<arangodb::rest::HttpCommTask> >::construct<arangodb::rest::HttpCommTask, arangodb::EventLoop&, arangodb::---Type <return> to continue, or q <return> to quit---
rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >, arangodb::ConnectionInfo, double&>(std::allocator<arangodb::rest::HttpCommTask>&, arangodb::rest::HttpCommTask*, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >&&, arangodb::ConnectionInfo&&, double&) (__p=<optimized out>, __a=...)
    at /usr/include/c++/5/bits/alloc_traits.h:530
#14 std::_Sp_counted_ptr_inplace<arangodb::rest::HttpCommTask, std::allocator<arangodb::rest::HttpCommTask>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >, arangodb::ConnectionInfo, double&>(std::allocator<arangodb::rest::HttpCommTask>, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >&&, arangodb::ConnectionInfo&&, double&) (__a=..., this=0x7f7edc2a2580)
    at /usr/include/c++/5/bits/shared_ptr_base.h:522
#15 std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<arangodb::rest::HttpCommTask, std::allocator<arangodb::rest::HttpCommTask>, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >, arangodb::ConnectionInfo, double&>(std::_Sp_make_shared_tag, arangodb::rest::HttpCommTask*, std::allocator<arangodb::rest::HttpCommTask> const&, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >&&, arangodb::ConnectionInfo&&, double&) (__a=..., this=<optimized out>) at /usr/include/c++/5/bits/shared_ptr_base.h:617
#16 std::__shared_ptr<arangodb::rest::HttpCommTask, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<arangodb::rest::HttpCommTask>, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >, arangodb::ConnectionInfo, double&>(std::_Sp_make_shared_tag, std::allocator<arangodb::rest::HttpCommTask> const&, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >&&, arangodb::ConnectionInfo&&, double&) (__a=..., __tag=..., 
    this=<optimized out>) at /usr/include/c++/5/bits/shared_ptr_base.h:1097
#17 std::shared_ptr<arangodb::rest::HttpCommTask>::shared_ptr<std::allocator<arangodb::rest::HttpCommTask>, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >, arangodb::ConnectionInfo, double&>(std::_Sp_make_shared_tag, std::allocator<arangodb::rest::HttpCommTask> const&, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >&&, arangodb::ConnectionInfo&&, double&) (__a=..., __tag=..., this=<optimized out>)
    at /usr/include/c++/5/bits/shared_ptr.h:319
#18 std::allocate_shared<arangodb::rest::HttpCommTask, std::allocator<arangodb::rest::HttpCommTask>, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >, arangodb::ConnectionInfo, double&>(std::allocator<arangodb::rest::HttpCommTask> const&, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >&&, arangodb::ConnectionInfo&&, double&) (__a=...) at /usr/include/c++/5/bits/shared_ptr.h:620
#19 std::make_shared<arangodb::rest::HttpCommTask, arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >, arangodb::ConnectionInfo, double&>(arangodb::EventLoop&, arangodb::rest::GeneralServer*&, std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >&&, arangodb::ConnectionInfo&&, double&) () at /usr/include/c++/5/bits/shared_ptr.h:636
#20 arangodb::rest::GeneralListenTask::handleConnected(std::unique_ptr<arangodb::Socket, std::default_delete<arangodb::Socket> >, arangodb::ConnectionInfo&&) (this=0x7f7ed429e100, socket=std::unique_ptr<arangodb::Socket> containing 0x0, 
    info=<unknown type in /home/steemann/Arango31NoAsan/build/bin/arangod, CU 0x3a7c930, DIE 0x3ad94a2>)
    at /home/steemann/Arango31NoAsan/arangod/GeneralServer/GeneralListenTask.cpp:65
#21 0x0000000000ade9b0 in arangodb::ListenTask::<lambda(const boost::system::error_code&)>::operator()(const boost::system::error_code &) const (
    __closure=0x7f7ed7ffbd50, ec=...) at /home/steemann/Arango31NoAsan/arangod/Scheduler/ListenTask.cpp:89
#22 0x0000000000ad5184 in std::function<void (boost::system::error_code const&)>::operator()(boost::system::error_code const&) const (
    __args#0=..., this=0x7f7ed7ffbd50) at /usr/include/c++/5/functional:2267
#23 boost::asio::detail::binder1<std::function<void (boost::system::error_code const&)>, boost::system::error_code>::operator()() (
    this=0x7f7ed7ffbd50) at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/detail/bind_handler.hpp:47
#24 boost::asio::asio_handler_invoke<boost::asio::detail::binder1<std::function<void (boost::system::error_code const&)>, boost::system::error_code> >(boost::asio::detail::binder1<std::function<void (boost::system::error_code const&)>, boost::system::error_code>&, ...) (function=...)
    at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/handler_invoke_hook.hpp:69
#25 boost_asio_handler_invoke_helpers::invoke<boost::asio::detail::binder1<std::function<void (boost::system::error_code const&)>, boost::system::error_code>, std::function<void (boost::system::error_code const&)> >(boost::asio::detail::binder1<std::function<void (boost::system::error_code const&)>, boost::system::error_code>&, std::function<void (boost::system::error_code const&)>&) (context=..., function=...)
    at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/detail/handler_invoke_helpers.hpp:37
#26 boost::asio::detail::reactive_socket_accept_op<boost::asio::basic_socket<boost::asio::ip::tcp, boost::asio::stream_socket_service<boost::asio::ip::tcp> >, boost::asio::ip::tcp, std::function<void (boost::system::error_code const&)> >::do_complete(boost::asio::detail::task_io_service*, boost::asio::detail::task_io_service_operation*, boost::system::error_code const&, unsigned long) (owner=0x7f7ee8e154e0, base=<optimized out>)
    at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/detail/reactive_socket_accept_op.hpp:123
#27 0x0000000000ad651e in boost::asio::detail::task_io_service_operation::complete (bytes_transferred=0, ec=..., owner=..., this=<optimized out>)
    at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/detail/task_io_service_operation.hpp:38
#28 boost::asio::detail::epoll_reactor::descriptor_state::do_complete (owner=0x7f7ee8e154e0, base=0x7f7ed3b943e0, ec=..., 
    bytes_transferred=<optimized out>) at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/detail/impl/epoll_reactor.ipp:651
#29 0x0000000000ae1fe6 in boost::asio::detail::task_io_service_operation::complete (bytes_transferred=1, ec=..., owner=..., this=0x7f7ed3b943e0)
    at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/detail/task_io_service_operation.hpp:38
#30 boost::asio::detail::task_io_service::do_run_one (ec=..., this_thread=..., lock=..., this=0x7f7ee8e154e0)
    at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/detail/impl/task_io_service.ipp:372
#31 boost::asio::detail::task_io_service::run_one (ec=..., this=0x7f7ee8e154e0)
    at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/detail/impl/task_io_service.ipp:170
#32 boost::asio::io_service::run_one (this=<optimized out>)
    at /home/steemann/Arango31NoAsan/3rdParty/boost/1.61.0/boost/asio/impl/io_service.ipp:72
#33 (anonymous namespace)::SchedulerThread::run (this=<optimized out>) at /home/steemann/Arango31NoAsan/arangod/Scheduler/Scheduler.cpp:105
#34 0x0000000000d5e6fc in arangodb::Thread::runMe (this=this@entry=0x7f7ee8d3e940) at /home/steemann/Arango31NoAsan/lib/Basics/Thread.cpp:374
#35 0x0000000000d5ed15 in arangodb::Thread::startThread (arg=0x7f7ee8d3e940) at /home/steemann/Arango31NoAsan/lib/Basics/Thread.cpp:83
#36 0x0000000000cebf1f in ThreadStarter (data=0x7f7edbd6f0e0) at /home/steemann/Arango31NoAsan/lib/Basics/threads-posix.cpp:67
#37 0x00007f7eea3be70a in start_thread (arg=0x7f7ed7ffd700) at pthread_create.c:333
#38 0x00007f7ee984c82d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

**Comment by jsteemann** [comment]
The problem seems to be that when an SSL connection is in the handshake phase and the underlying peer connection gets closed at this stage, the handshake process will not finish. Thus the scheduler thread will continue with high CPU usage, and shutdown does not work because the thread does not return.
A preliminary fix for this has been pushed in 3.1 and devel. The fix will abort long-running handshakes so that threads can recover from this state. 

A better fix for a future release will be to make the handshake operation asynchronous and not block the scheduler thread and not use any CPU in case nothing happens. But this has still to be implemented.

**Comment by jsteemann** [comment]
3.1.5 has been released. Closing as fixed.
